src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForUnionAlgorithms.java

```diff
@@ -0,0 +1,222 @@
+			assertEquals( 1, b.size() );
+			if (b.get(x).getURI().equals("http://example.org/x1"))
+				b1Found = true;
+			else if (b.get(x).getURI().equals("http://example.org/x2"))
+				b2Found = true;
+			else if ((b.get(x).getURI().equals("http://example.org/x3")))
+				b3Found = true;
+			else {
+				fail( "Unexpected URI for ?x: " + b.get(x).getURI() );
+			}
+		}
+		assertTrue(b1Found);
+		assertTrue(b2Found);
+		assertTrue(b3Found);
+	}
+	
+	protected void _testUnboundUnion() {
+		final Var x = Var.alloc("x");
+		final Var y = Var.alloc("y");
+		
+		final Node x1 = NodeFactory.createURI("http://example.org/x1");
+		final Node x2 = NodeFactory.createURI("http://example.org/x2");
+		final Node x3 = NodeFactory.createURI("http://example.org/x3");
+		final Node y1 = NodeFactory.createURI("http://example.org/y1");
+		final Node y2 = NodeFactory.createURI("http://example.org/y2");
+		final Node y3 = NodeFactory.createURI("http://example.org/y3");
+		
+		final GenericIntermediateResultBlockImpl input1 = new GenericIntermediateResultBlockImpl();
+		input1.add( SolutionMappingUtils.createSolutionMapping(x, x1) );
+		input1.add( SolutionMappingUtils.createSolutionMapping(x, x3, y, y3) );
+		
+		final GenericIntermediateResultBlockImpl input2 = new GenericIntermediateResultBlockImpl();
+		input2.add( SolutionMappingUtils.createSolutionMapping(y, y1) );
+		input2.add( SolutionMappingUtils.createSolutionMapping(y, y2, x, x2) );
+		
+		final Iterator<SolutionMapping> it = runTest(input1, input2);
+		
+		final List<Binding> result = new ArrayList<>();
+	
+		assertTrue( it.hasNext() );
+		result.add( it.next().asJenaBinding() );
+	
+		assertTrue( it.hasNext() );
+		result.add( it.next().asJenaBinding() );
+	
+		assertTrue( it.hasNext() );
+		result.add( it.next().asJenaBinding() );
+		
+		assertTrue( it.hasNext() );
+		result.add( it.next().asJenaBinding() );
+	
+		assertFalse( it.hasNext() );
+	
+		boolean b1Found = false;
+		boolean b2Found = false;
+		boolean b3Found = false;
+		boolean b4Found = false;
+		
+		for ( final Binding b : result ) {
+			if (b.get(x) != null) {
+				if (b.get(x).getURI().equals("http://example.org/x1")) {
+					assertEquals(1, b.size());
+					b1Found = true;
+				}
+				else if (b.get(x).getURI().equals("http://example.org/x2")) {
+					assertEquals(2, b.size());
+					assertNotEquals(null, b.get(y));
+					assertEquals("http://example.org/y2", b.get(y).getURI());
+					b2Found = true;
+				}
+				else if (b.get(x).getURI().equals("http://example.org/x3")) {
+					assertEquals(2, b.size());
+					assertNotEquals(null, b.get(y));
+					assertEquals("http://example.org/y3", b.get(y).getURI());
+					b3Found = true;
+				}
+			}
+			else if (b.get(y) != null && b.get(y).getURI().equals("http://example.org/y1")) {
+				assertEquals(1, b.size());
+				b4Found = true;
+			}
+		}
+		assertTrue(b1Found);
+		assertTrue(b2Found);
+		assertTrue(b3Found);
+		assertTrue(b4Found);
+	}
+	
+	protected void _testKeepUnionDuplicates() {

```

Comment:
Can you please add another such unit test where the duplicates come from the same input. Just to make sure that implementations get this option right as well.

Reply:
Thanks for adding the extra unit test !